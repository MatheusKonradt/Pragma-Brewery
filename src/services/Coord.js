
class Coord {
  /**
   * @param {number} x
   * @param {number} y
   */
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  /**
   * @param {CoordSpace} val
   * @return {Coord}
   */
  setSpace(val) {
    this.space = val;
    return this;
  }

  /**
   * @param {CoordSpace} toSpace
   * @return {Coord}
   */
  transformSpace(toSpace) {
    const coord = this.space.transformCoordSpace(this, toSpace);
    this.x = coord.x;
    this.y = coord.y;
    this.setSpace(toSpace);
    return this;
  }

  /**
   * @return {Coord}
   */
  invertYAxis() {
    this.y = this.space.yInterval.b - this.y;
    return this;
  }
}

export default Coord;
