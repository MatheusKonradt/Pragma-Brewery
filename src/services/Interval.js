class Interval {
  /**
   * @param {number} a
   * @param {number} b
   */
  constructor(a, b) {
    this.a = a;
    this.b = b;
    if (a > b) throw new Error('invalid interval.');
  }

  /**
   * @return {number}
   */
  getSize() {
    return this.b - this.a;
  }

  /**
   * @param {number} number
   * @param {Interval} interval
   * @return {number}
   */
  scaleNumberToOtherInterval(number, interval) {
    const ratio = interval.getSize() / this.getSize();
    return (ratio * number) - (ratio * this.a) + interval.a;
  }

  /**
   * @param {number} step
   * @param {Function} cb
   * @return {Array<any>}
   */
  map(step, cb) {
    const values = [];
    for(let i=this.a ; i<this.b ; i+= step) values.push(i);
    return values.map(cb);
  }
}

export default Interval;