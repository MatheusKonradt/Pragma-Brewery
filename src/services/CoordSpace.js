import Coord from "./Coord";

class CoordSpace {
  /**
   * @param {Interval} xInterval
   * @param {Interval} yInterval
   */
  constructor(xInterval, yInterval) {
    this.yInterval = yInterval;
    this.xInterval = xInterval;
  }

  /**
   * @param {Coord} coord
   * @param {CoordSpace} coordSpace
   * @return {Coord}
   */
  transformCoordSpace(coord, coordSpace) {
    const x = this.xInterval.scaleNumberToOtherInterval(coord.x, coordSpace.xInterval);
    const y = this.yInterval.scaleNumberToOtherInterval(coord.y, coordSpace.yInterval);
    return new Coord(x, y);
  }
}

export default CoordSpace;
