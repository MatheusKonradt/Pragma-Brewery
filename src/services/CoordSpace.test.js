import CoordSpace from './CoordSpace';
import Interval from './Interval';
import Coord from './Coord';

test('Should transform coordinates from a space to another', () => {
  const coordSpaceA = new CoordSpace(new Interval(0, 5), new Interval(0, 20));
  const coordSpaceB = new CoordSpace(new Interval(0, 10), new Interval(10, 15));
  const coord = new Coord(4, 15);
  const transformedCoord = coordSpaceA.transformCoordSpace(coord, coordSpaceB);
  expect(transformedCoord).toMatchSnapshot();
});
