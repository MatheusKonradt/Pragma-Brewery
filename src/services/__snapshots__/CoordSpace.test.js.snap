// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should transform coordinates from a space to another 1`] = `
Coord {
  "x": 8,
  "y": 13.75,
}
`;
